### ISSUE TITLE
The restore snapshot script fails if a VM controls the VM host

### ISSUE THREAD

**Issue by jafz** [AITGmbH/AIT.BuildSuite.HyperV#4]
I created a VM called 'VMHOST-CTRL1' on a VM-Host, so that the host does not need to run an agent and is purely a host.
The Restore Snapshot script runs 'Enable-VMEventing -force' at the end. However, that requires the hyper-v feature to be installed, which is not the case if a machine is simply able to remotely manage a vm host.
I was forced to use nested virtualization and install the hyper-v feature on the VM.

Exception from the log:

2019-04-11T12:34:55.2409828Z Restoring snapshots has been finished for all VM(s).
2019-04-11T12:34:55.2409935Z 
2019-04-11T12:34:55.2409991Z 
2019-04-11T12:34:55.2410088Z (Re-)Enable Hyper-V cmdlet caching.
2019-04-11T12:34:55.2410156Z 
2019-04-11T12:34:55.2410212Z 
2019-04-11T12:34:55.4325466Z ##[error]Microsoft.HyperV.PowerShell.VirtualizationException: The Hyper-V Management Tools could not access an expected WMI class on computer 'VMHOST-CTRL1'. This may indicate that the Hyper-V Platform is not installed on the computer or that the version of the Hyper-V Platform is incompatible with these management tools. ---> Microsoft.Virtualization.Client.Management.ServerConnectionException: The Hyper-V Management Tools could not access an expected WMI class on computer 'VMHOST-CTRL1'. This may indicate that the Hyper-V Platform is not installed on the computer or that the version of the Hyper-V Platform is incompatible with these management tools. ---> Microsoft.Management.Infrastructure.CimException: Invalid class 
2019-04-11T12:34:55.4336389Z    at Microsoft.Management.Infrastructure.Internal.Operations.CimSyncEnumeratorBase`1.MoveNext()
2019-04-11T12:34:55.4336702Z    at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
2019-04-11T12:34:55.4336947Z    at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
2019-04-11T12:34:55.4337168Z    at Microsoft.Virtualization.Client.Management.Server.LoadOSInfo()
2019-04-11T12:34:55.4337377Z    --- End of inner exception stack trace ---
2019-04-11T12:34:55.4337593Z    at Microsoft.Virtualization.Client.Management.Server.LoadOSInfo()
2019-04-11T12:34:55.4337833Z    at Microsoft.HyperV.PowerShell.VirtualizationObjectLocator.PrepareServerConnections(Server server, Boolean allowDownlevelServers)
2019-04-11T12:34:55.4338093Z    at Microsoft.HyperV.PowerShell.Commands.ParameterResolvers.GetServers(IServerParameters cmdletParameters, IOperationWatcher operationWatcher)
2019-04-11T12:34:55.4338375Z    at Microsoft.HyperV.PowerShell.Commands.EnableVMEventing.PerformOperation(IOperationWatcher operationWatcher)
2019-04-11T12:34:55.4338635Z    at Microsoft.HyperV.PowerShell.Commands.VirtualizationCmdletBase.PerformOperationWithLogging(IOperationWatcher operationWatcher)
2019-04-11T12:34:55.4338872Z    --- End of inner exception stack trace ---

**Comment by norschel** [comment]
@jafz , can you tell me more about your scenario? Do you use the Azure DevOps Extension published in the Marketplace or directly the script?

The script itself (which is called in the extension) doesn't really need much, just the HyperV management tools and firewall releases. It can happen that the calls from your VM are blocked due to missing firewall port rules. It is also important that the management tools are compatible between VM and host. This is explicitly listed in the Extension Readme file (see requirements section). Your output also points in this direction ... see "The Hyper-V Management Tools could not access an expected WMI class on computer 'VMHOST-CTRL1'. This may indicate that the Hyper-V Platform is not installed on the computer or that the version of the Hyper-V Platform is incompatible with these management tools." 
Please also check the hostname. It looks to me like you have entered the management VM as your host name. But here the Hyper-V hostname has to be entered as hostname.

**Comment by jafz** [comment]
Hi norschel
thanks for the fast reply.

I use the marketplace extension in our local TFS installation, but I opened the script to see what's happening.
The VMHost ist a windows 2019 core installation, the 'VMHOST-CTRL1' is a VM with Win2019 (desktop experience, including the hyper-v powershell commands feature) on that VMHost, while the RegressionTestVm is the VM where I want my snapshot applied.
The setup is like this because I do not want to install TFS agents on the VMHost.

win2019->win2019 seems to me that both should be compatible. Both machines are also in the same domain. The reset of the VM snapshot actually works. What does not work is **the command 'Enable-VMEventing -force'. And that command, if I read that script correctly, is executed in the function that prints '(Re-)Enable Hyper-V cmdlet caching.' to the log, it is executed _without_ the parameter -ComputerName.** This makes that command execute on the VM itself, not on the remote VM Host, and that is what fails.
I have not tried executing that command on a win2012 or win2016 machine yet (those machines would also be simply VMs without hosting capability).

As you can see in the screenshot: VMHOST-CTRL1 is not even mentioned in my configuration. It is simply the VM that the TFS agent runs on. VMHost3 is the host.

Bottom line: the script executes enable-vmeventing -force on a non-host, that fails, and that exception isn't caught, nor is the eventing function told to execute on a remote host.

![image](https://user-images.githubusercontent.com/4342877/56082996-b8dd0380-5e1f-11e9-846d-045bc8cd16bb.png)

**Comment by norschel** [comment]
Now I understand the problem so far. There actually seems to be an argument missing (computer name). Exciting why it worked for me without it. My Hyper-V server is also only called remotely. I have already added the correction to the current sources. 
Since the current Master Branch has a few more changes ( switching PowershellHandler to PowerShellHandler3) and an intensive testing is still pending, the fix cannot go live yet.

Can you try the new preview version? 
[<= TFS 2018](https://dev.azure.com/ait-public/_apis/resources/Containers/2764026?itemPath=Extension%2FAIT.BuildSuite.HyperV__build_and_release_preview.vsix
)
[>= Azure DevOps 2019](https://dev.azure.com/ait-public/_apis/resources/Containers/2764026?itemPath=Extension%2FAIT.BuildSuite.HyperV_azurepipelines_preview.vsix)

Win 2019 wasn't tested yet, because the extension was released before Win 2019. After your test we can certainly add Win 2019 support. It should work today, but better test it explicitly again.

**Comment by jafz** [comment]
I wasn't able to get the preview version, as those links gave me an unauthorized.
But I put the changes of commit 56d89fb to the script, and now it seems to work, even though the machine doesn't have hyper-v installed.
Thanks.
